File: Assets/ToLua/Lua/user/ui/item/controller/item_controller.lua
local t = {}
local name = 'item_controller'

require 'item_pb'

local function Start()
  netmanager.RegisterProtocol(MSG.GetAllItemResp,t.GetAllItemResp)
  netmanager.RegisterProtocol(MSG.ItemUpdateResp,t.ItemUpdateResp)
  netmanager.RegisterProtocol(MSG.ExpPotionResp,t.ExpPotionResp)
  netmanager.RegisterProtocol(MSG.OpenGiftBagResp,t.OpenGiftBagResp)
  gamemanager.RegisterCtrl(name,t)
end

--------------------req-----------------
--经验药水
function t.ExpPotionReq(heroid, potionNo, num)
  local req = item_pb.ExpPotionReq()
  req.heroId = heroid
  req.potionNo = potionNo
  req.num = num
  netmanager.SendProtocol(MSG.ExpPotionReq, req)
end

-------------------resp----------------------------------
function t.GetAllItemResp()
  
  local resp = item_pb.GetAllItemResp()
  resp:ParseFromString(netmanager.GetProtocolData())
  local item_info = require('ui/item/model/item_info')
  local item_model = gamemanager.GetModel('item_model')
  
  for k,v in ipairs(resp.items) do
     item_model.AddItemInfo(item_info.NewByItemProtoData(v), false)
  end
  Observers.Facade.Instance:SendNotification('LOBBY2CLIENT_GET_ALL_ITEM_RESP_handler')
end

function t.ItemUpdateResp()
  local resp = item_pb.ItemUpdateResp()
  resp:ParseFromString(netmanager.GetProtocolData())
  local item_info = require('ui/item/model/item_info')
  local item_model = gamemanager.GetModel('item_model')
  --add
  for k,v in ipairs(resp.addItems) do 
    item_model.AddItemInfo(item_info.NewByItemProtoData(v), true)
    --print('add item..............',v.id)
  end
  --remove
  local tempInfo = nil
  for k,v in ipairs(resp.delItems) do
    tempInfo = item_model.GetItemInfoByInstanceID(v)
    item_model.RemoveItemInfo(v)
    TalkingDataController.instance:TDGAItemOnUse(tempInfo.itemData.id,tempInfo:Count(),BaseResType.Item)
    --print('remove item..............',v)
  end
  --update
  local tempCount = 0
  for k,v in ipairs(resp.updateItems) do
    local info = item_model.GetItemInfoByInstanceID(v.id)
    tempCount = info:Count()
    info:Update(v)
    --print('update item...................',info.instanceId,info.count)
    if tempCount > info:Count() then
      TalkingDataController.instance:TDGAItemOnUse(info.itemData.id,tempCount-info:Count(),BaseResType.Item)
    end
  end
  item_model.updateItemInfoListByProtocol()
  Observers.Facade.Instance:SendNotification('LOBBY2CLIENT_ITEM_UPDATE_RESP_handler')
  gamemanager.GetModel('hero_model').CheckHasAdvanceBreakthroughHeroByRedPoint()
end

function t.ExpPotionResp()
  local resp = item_pb.ExpPotionResp()
  resp:ParseFromString(netmanager.GetProtocolData())
  local item_model = gamemanager.GetModel('item_model')
  item_model.updateExpPotionByProtocol(resp.potionNo)
end

function t.OpenGiftBagReq(itemid)
  local req = item_pb.OpenGiftBagReq()
  req.giftBagNo = itemid
  netmanager.SendProtocol(MSG.OpenGiftBagReq, req)
end

function t.OpenGiftBagResp()
  local resp = item_pb.OpenGiftBagResp()
  resp:ParseFromString(netmanager.GetProtocolData())
  local dataList = {}
  for k,v in ipairs(resp.dropItems) do
    dataList[k] = require('ui/game/model/game_res_data').New(v.itemType,v.itemNo,v.itemNum,v.heroStar)
  end
  require('ui/tips/view/common_reward_tips_view').Create(dataList)
end

Start()

return t

File: Assets/Code/Logic/Item/Controller/ItemController.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using Logic.Protocol.Model;
using Logic.Item.Model;
using LuaInterface;

namespace Logic.Item.Controller
{
	public class ItemController : SingletonMono<ItemController> 
	{
		
		void Awake()
		{
			instance = this;
		}

		void Start()
		{
			//Observers.Facade.Instance.RegisterObserver(((int)MSG.GetAllItemResp).ToString(),LOBBY2CLIENT_GET_ALL_ITEM_RESP_handler);
			//Observers.Facade.Instance.RegisterObserver(((int)MSG.ItemUpdateResp).ToString(),LOBBY2CLIENT_ITEM_UPDATE_RESP_handler);
			//Observers.Facade.Instance.RegisterObserver(((int)MSG.ExpPotionResp).ToString(),LOBBY2CLIENT_ExpPotion_RESP_handler);
			Observers.Facade.Instance.RegisterObserver("LOBBY2CLIENT_GET_ALL_ITEM_RESP_handler",LOBBY2CLIENT_GET_ALL_ITEM_RESP_handler);
			Observers.Facade.Instance.RegisterObserver("LOBBY2CLIENT_ITEM_UPDATE_RESP_handler",LOBBY2CLIENT_ITEM_UPDATE_RESP_handler);
			Observers.Facade.Instance.RegisterObserver(((int)MSG.BuySweepCouponsResp).ToString(),LOBBY2CLIENT_BuySweepCouponsResp_RESP_handler);
			
		}
		void OnDestroy()
		{
			if (Observers.Facade.Instance != null)
			{
				//Observers.Facade.Instance.RemoveObserver(((int)MSG.GetAllItemResp).ToString(),LOBBY2CLIENT_GET_ALL_ITEM_RESP_handler);
				//Observers.Facade.Instance.RemoveObserver(((int)MSG.ItemUpdateResp).ToString(),LOBBY2CLIENT_ITEM_UPDATE_RESP_handler);
				//Observers.Facade.Instance.RemoveObserver(((int)MSG.ExpPotionResp).ToString(),LOBBY2CLIENT_ExpPotion_RESP_handler);
				Observers.Facade.Instance.RemoveObserver("LOBBY2CLIENT_GET_ALL_ITEM_RESP_handler",LOBBY2CLIENT_GET_ALL_ITEM_RESP_handler);
				Observers.Facade.Instance.RemoveObserver("LOBBY2CLIENT_ITEM_UPDATE_RESP_handler",LOBBY2CLIENT_ITEM_UPDATE_RESP_handler);
				Observers.Facade.Instance.RemoveObserver(((int)MSG.BuySweepCouponsResp).ToString(),LOBBY2CLIENT_BuySweepCouponsResp_RESP_handler);
				
			}
		}

		public void CLIENT2LOBBY_ExpPotion_REQ(int heroid,int potionNo,int num)
		{
			ExpPotionReq req = new ExpPotionReq();
			req.heroId = heroid;
			req.potionNo = potionNo;
			req.num = num;
			Protocol.ProtocolProxy.instance.SendProtocol(req);
		}
		public void CLIENT2LOBBY_BuySweepCouponsReq(int num)
		{
			IntProto req = new IntProto();
			req.value = num;
			Protocol.ProtocolProxy.instance.SendProtocol((int)MSG.BuySweepCouponsReq, req);
		}
		private bool LOBBY2CLIENT_GET_ALL_ITEM_RESP_handler(Observers.Interfaces.INotification note)
		{
//			GetAllItemResp resp = note.Body as GetAllItemResp;
//			int count = resp.items.Count;
//			for(int i = 0;i<count;i++){
//				ItemInfo item = new ItemInfo(resp.items[i]);
//				ItemProxy.instance.AddItemInfo(item);
//			}
			LuaTable itemModel = (LuaTable)LuaScriptMgr.Instance.CallLuaFunction("gamemanager.GetModel","item_model")[0];
			LuaTable itemsTable = (LuaTable)itemModel["itemsTable"];
			foreach(DictionaryEntry kvp in itemsTable.ToDictTable())
			{
				LuaTable itemInfoTable = (LuaTable)kvp.Value;
				int id = itemInfoTable["instanceId"].ToString().ToInt32();
				int count = itemInfoTable["count"].ToString().ToInt32();
				LuaTable itemDataTable = (LuaTable)itemInfoTable["itemData"];
				if(itemDataTable != null)
				{
					int modelId = itemDataTable["id"].ToString().ToInt32();
					ItemProxy.instance.AddItemInfo(new ItemInfo(id,modelId,count));
				}
			}
			Logic.UI.RandomGift.Model.RandomGiftProxy.instance.UpdateRandomGift();
			return true;
		}
		private bool LOBBY2CLIENT_ITEM_UPDATE_RESP_handler(Observers.Interfaces.INotification note)
		{
//			ItemUpdateResp resp = note.Body as ItemUpdateResp;
//			//add
//			int addCount = resp.addItems.Count;
//			for(int i = 0;i<addCount;i++)
//			{
//				ItemInfo item = new ItemInfo(resp.addItems[i]);
//				ItemProxy.instance.AddItemInfo(item);
//			}
//			//remove
//			int removeCount = resp.delItems.Count;
//			for(int i = 0;i<removeCount;i++)
//			{
//				ItemProxy.instance.RemoveItemInfo(resp.delItems[i]);
//			}
//			//update
//			int updateCount = resp.updateItems.Count;
//			for(int i = 0;i<updateCount;i++)
//			{
//				ItemProxy.instance.UpdateItemInfo(resp.updateItems[i]);
//			}
			ItemProxy.instance.GetAllItemInfoDictioary().Clear();

			LuaTable itemModel = (LuaTable)LuaScriptMgr.Instance.CallLuaFunction("gamemanager.GetModel","item_model")[0];
			LuaTable itemsTable = (LuaTable)itemModel["itemsTable"];
			foreach(DictionaryEntry kvp in itemsTable.ToDictTable())
			{
				LuaTable itemInfoTable = (LuaTable)kvp.Value;
				int id = itemInfoTable["instanceId"].ToString().ToInt32();
				int count = itemInfoTable["count"].ToString().ToInt32();
				LuaTable itemDataTable = (LuaTable)itemInfoTable["itemData"];
				if(itemDataTable != null)
				{
					int modelId = itemDataTable["id"].ToString().ToInt32();
					ItemProxy.instance.AddItemInfo(new ItemInfo(id,modelId,count));
				}
			}

			ItemProxy.instance.OnItemInfoListUpdate();

			//if(addCount != 0 || removeCount != 0)
				Logic.UI.RandomGift.Model.RandomGiftProxy.instance.UpdateRandomGift();
			return true;
		}
		private bool LOBBY2CLIENT_ExpPotion_RESP_handler(Observers.Interfaces.INotification note)
		{
			return true;
		}
		private bool LOBBY2CLIENT_BuySweepCouponsResp_RESP_handler(Observers.Interfaces.INotification note)
		{
			Logic.UI.DungeonDetail.Model.DungeonDetailProxy.instance.OnBuySweepCouponsSuccess();
			return true;
		}
	}
}

